shell/platform/android/io/flutter/app/FlutterPluginRegistry.java

```diff
@@ -0,0 +1,172 @@
+// Copyright 2017 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+package io.flutter.app;
+
+import android.app.Activity;
+import android.content.Context;
+import android.util.Log;
+import io.flutter.plugin.common.*;
+import io.flutter.view.FlutterNativeView;
+import io.flutter.view.FlutterView;
+import io.flutter.view.TextureRegistry;
+
+import java.util.ArrayList;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+
+public class FlutterPluginRegistry implements PluginRegistry {
+    private static final String TAG = "FlutterPluginRegistry";
+
+    private Activity mActivity;
+    private Context mAppContext;
+    private FlutterNativeView mNativeView;
+    private FlutterView mFlutterView;
+
+    private final Map<String, Object> mPluginMap = new LinkedHashMap<>(0);
+    private final List<RequestPermissionResultListener> mRequestPermissionResultListeners = new ArrayList<>(0);
+    private final List<ActivityResultListener> mActivityResultListeners = new ArrayList<>(0);
+    private final List<NewIntentListener> mNewIntentListeners = new ArrayList<>(0);
+    private final List<UserLeaveHintListener> mUserLeaveHintListeners = new ArrayList<>(0);
+    private final List<ViewDestroyListener> mViewDestroyListeners = new ArrayList<>(0);
+
+    public FlutterPluginRegistry(FlutterNativeView nativeView, Context context) {
+        mNativeView = nativeView;
+        mAppContext = context;
+    }
+
+    @Override
+    public boolean hasPlugin(String key) {
+        return mPluginMap.containsKey(key);
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public <T> T valuePublishedByPlugin(String pluginKey) {
+        return (T) mPluginMap.get(pluginKey);
+    }
+
+    @Override
+    public Registrar registrarFor(String pluginKey) {
+        if (mPluginMap.containsKey(pluginKey)) {
+            throw new IllegalStateException("Plugin key " + pluginKey + " is already in use");
+        }
+        mPluginMap.put(pluginKey, null);
+        return new FlutterRegistrar(pluginKey);
+    }
+
+    public void attach(FlutterView flutterView, Activity activity) {
+        mFlutterView = flutterView;
+        mActivity = activity;
+    }
+
+    public void detach() {
+        mFlutterView = null;
+        mActivity = null;
+    }
+
+    public List<RequestPermissionResultListener> getRequestPermissionResultListeners() {
+        return mRequestPermissionResultListeners;
```

Comment:
1. These methods now expose mutable objects that are otherwise managed by the registry.
1. They adobt the java beans method naming convention (adding a `get` prefix to every getter) which the rest of the API does not.
1. They require callers to ask for information rather than help.

I suggest refactoring this to make the registry class implement the `XxxListener` interfaces:
```java
class FlutterPluginRegistry implements PluginRegistry, XxxListener, ... {
  @Override
  public boolean onXxx(X x) {
    for (XxxListener listener : mXxxListeners) {
      if (listener.onXxx(x)) {
        return true;
      }
    }
    return false;
  }
  ...
}
```
Then the client code in `FlutterActivityDelegate` can just ask for help:
```java
   @Override
   public boolean onXxx(X x) {
      return flutterView.getPluginRegistry().onXxx(x);
   }
 ```

Reply:
Done